File: engine/yai-engine/src/projection.rs
use crate::journal::Journal;
use crate::record::RecordKind;
use std::collections::BTreeSet;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ProjectionSummary {
    pub consumer: String,
    pub case_ref: String,
    pub summary: String,
    pub source_record_count: usize,
    pub receipt_count: usize,
    pub subject_count: usize,
    pub decision_count: usize,
    pub policy_rule_count: usize,
    pub gate_count: usize,
    pub obligation_count: usize,
    pub receipt_requirement_count: usize,
    pub filesystem_receipt_count: usize,
    pub subject_state_count: usize,
    pub effect_count: usize,
    pub graph_edge_count: usize,
    pub reconstruction_count: usize,
    pub memory_candidate_count: usize,
    pub operational_memory_candidate_count: usize,
    pub decision_memory_candidate_count: usize,
    pub subject_memory_candidate_count: usize,
    pub error_memory_candidate_count: usize,
    pub recovery_memory_candidate_count: usize,
    pub divergence_count: usize,
    pub reconciliation_count: usize,
    pub critical_divergence_count: usize,
    pub warning_divergence_count: usize,
    pub projection_request_count: usize,
    pub projection_result_count: usize,
    pub operator_projection_count: usize,
    pub model_projection_count: usize,
    pub audit_projection_count: usize,
    pub limited_projection_count: usize,
    pub query_result_count: usize,
}

impl ProjectionSummary {
    pub fn from_journal(consumer: impl Into<String>, journal: &Journal) -> Self {
        let consumer = consumer.into();
        let source_record_count = journal.count();
        let case_ref = journal
            .records()
            .iter()
            .find(|record| !record.case_ref.is_empty())
            .map(|record| record.case_ref.clone())
            .unwrap_or_default();
        let receipt_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Receipt)
            .count();
        let decision_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Decision)
            .count();
        let policy_rule_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::PolicyRule)
            .count();
        let gate_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::GateResult)
            .count();
        let obligation_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Obligation)
            .count();
        let receipt_requirement_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::ReceiptRequirement)
            .count();
        let filesystem_receipt_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::FilesystemReceipt)
            .count();
        let subject_state_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::SubjectState)
            .count();
        let effect_count = journal
            .records()
            .iter()
            .filter(|record| {
                matches!(
                    record.kind,
                    RecordKind::Receipt | RecordKind::EffectReceipt | RecordKind::FilesystemReceipt
                )
            })
            .count();
        let graph_edge_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::GraphEdge)
            .count();
        let reconstruction_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Reconstruction)
            .count();
        let memory_records = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::MemoryCandidate)
            .collect::<Vec<_>>();
        let memory_candidate_count = memory_records.len();
        let operational_memory_candidate_count = memory_records
            .iter()
            .filter(|record| record.summary.contains("memory:operational"))
            .count();
        let decision_memory_candidate_count = memory_records
            .iter()
            .filter(|record| record.summary.contains("memory:decision"))
            .count();
        let subject_memory_candidate_count = memory_records
            .iter()
            .filter(|record| record.summary.contains("memory:subject"))
            .count();
        let error_memory_candidate_count = memory_records
            .iter()
            .filter(|record| record.summary.contains("memory:error"))
            .count();
        let recovery_memory_candidate_count = memory_records
            .iter()
            .filter(|record| record.summary.contains("memory:recovery"))
            .count();
        let subject_count = journal
            .records()
            .iter()
            .filter(|record| !record.subject_ref.is_empty() && record.subject_ref != "subject:none")
            .map(|record| record.subject_ref.as_str())
            .collect::<BTreeSet<_>>()
            .len();
        let divergence_records = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Divergence)
            .collect::<Vec<_>>();
        let divergence_count = divergence_records.len();
        let reconciliation_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::Reconciliation)
            .count();
        let critical_divergence_count = divergence_records
            .iter()
            .filter(|record| record.summary.contains("severity:critical"))
            .count();
        let warning_divergence_count = divergence_records
            .iter()
            .filter(|record| record.summary.contains("severity:warning"))
            .count();
        let projection_request_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::ProjectionRequest)
            .count();
        let projection_result_records = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::ProjectionResult)
            .collect::<Vec<_>>();
        let projection_result_count = projection_result_records.len();
        let operator_projection_count = projection_result_records
            .iter()
            .filter(|record| record.summary.contains("consumer:operator"))
            .count();
        let model_projection_count = projection_result_records
            .iter()
            .filter(|record| record.summary.contains("consumer:model"))
            .count();
        let audit_projection_count = projection_result_records
            .iter()
            .filter(|record| record.summary.contains("consumer:audit"))
            .count();
        let limited_projection_count = projection_result_records
            .iter()
            .filter(|record| {
                record.summary.contains("redaction:summary_only")
                    || record.summary.contains("redaction:refs_only")
                    || record.summary.contains("redaction:redacted")
                    || record.summary.contains("redaction:blocked")
            })
            .count();
        let query_result_count = journal
            .records()
            .iter()
            .filter(|record| record.kind == RecordKind::QueryResult)
            .count();
        Self {
            summary: format!(
                "projection:{consumer} records:{source_record_count} decisions:{decision_count} rules:{policy_rule_count} gates:{gate_count} obligations:{obligation_count} receipt_requirements:{receipt_requirement_count} filesystem_receipts:{filesystem_receipt_count} subject_states:{subject_state_count} effects:{effect_count} graph_edges:{graph_edge_count} reconstructions:{reconstruction_count} memory_candidates:{memory_candidate_count} divergences:{divergence_count} reconciliations:{reconciliation_count} projection_requests:{projection_request_count} projection_results:{projection_result_count} query_results:{query_result_count}"
            ),
            consumer,
            case_ref,
            source_record_count,
            receipt_count,
            subject_count,
            decision_count,
            policy_rule_count,
            gate_count,
            obligation_count,
            receipt_requirement_count,
            filesystem_receipt_count,
            subject_state_count,
            effect_count,
            graph_edge_count,
            reconstruction_count,
            memory_candidate_count,
            operational_memory_candidate_count,
            decision_memory_candidate_count,
            subject_memory_candidate_count,
            error_memory_candidate_count,
            recovery_memory_candidate_count,
            divergence_count,
            reconciliation_count,
            critical_divergence_count,
            warning_divergence_count,
            projection_request_count,
            projection_result_count,
            operator_projection_count,
            model_projection_count,
            audit_projection_count,
            limited_projection_count,
            query_result_count,
        }
    }
}

File: system/projection/projection.c
#include "yai/projection/projection.h"

#include <stdio.h>
#include <string.h>

static int summary_has_memory_kind(const yai_store_record_t *record,
                                   const char *kind) {
    char pattern[64];

    if (record == 0 || kind == 0) {
        return 0;
    }
    (void)snprintf(pattern, sizeof(pattern), "memory:%s", kind);
    return strstr(record->summary, pattern) != 0;
}

yai_status_t yai_projection_build(const char *projection_id,
                                  const yai_case_ref_t *case_ref,
                                  yai_projection_consumer_t consumer_kind,
                                  const yai_journal_t *journal,
                                  yai_projection_t *projection) {
    size_t index = 0;

    if (projection == 0 || projection_id == 0 || projection_id[0] == '\0' ||
        yai_case_ref_is_empty(case_ref) || journal == 0) {
        return YAI_ERR_INVALID;
    }

    yai_id_set(&projection->projection_id, projection_id);
    projection->case_ref = *case_ref;
    projection->consumer_kind = consumer_kind;
    projection->source_record_count = yai_journal_count(journal);
    projection->decision_count = 0;
    projection->policy_rule_count = 0;
    projection->gate_count = 0;
    projection->obligation_count = 0;
    projection->receipt_requirement_count = 0;
    projection->filesystem_receipt_count = 0;
    projection->subject_state_count = 0;
    projection->effect_count = 0;
    projection->graph_edge_count = 0;
    projection->reconstruction_count = 0;
    projection->memory_candidate_count = 0;
    projection->operational_memory_candidate_count = 0;
    projection->decision_memory_candidate_count = 0;
    projection->subject_memory_candidate_count = 0;
    projection->error_memory_candidate_count = 0;
    projection->recovery_memory_candidate_count = 0;
    projection->divergence_count = 0;
    projection->reconciliation_count = 0;
    projection->critical_divergence_count = 0;
    projection->warning_divergence_count = 0;
    projection->projection_request_count = 0;
    projection->projection_result_count = 0;
    projection->operator_projection_count = 0;
    projection->model_projection_count = 0;
    projection->audit_projection_count = 0;
    projection->limited_projection_count = 0;
    projection->query_result_count = 0;
    projection->projection_rule_count = 0;
    projection->authority_scope_count = 0;
    projection->model_interpretation_count = 0;
    projection->case_domain_count = 0;
    projection->case_attachment_count = 0;
    projection->case_binding_count = 0;
    projection->interaction_thread_count = 0;
    projection->interaction_turn_count = 0;
    projection->participant_view_frame_count = 0;

    for (index = 0; index < yai_journal_count(journal); index += 1) {
        const yai_store_record_t *record = yai_journal_get(journal, index);
        if (record == 0) {
            continue;
        }
        if (record->record_kind == YAI_RECORD_DECISION) {
            projection->decision_count += 1;
        } else if (record->record_kind == YAI_RECORD_POLICY_RULE) {
            projection->policy_rule_count += 1;
        } else if (record->record_kind == YAI_RECORD_GATE_RESULT) {
            projection->gate_count += 1;
        } else if (record->record_kind == YAI_RECORD_OBLIGATION) {
            projection->obligation_count += 1;
        } else if (record->record_kind == YAI_RECORD_RECEIPT_REQUIREMENT) {
            projection->receipt_requirement_count += 1;
        } else if (record->record_kind == YAI_RECORD_FILESYSTEM_RECEIPT) {
            projection->filesystem_receipt_count += 1;
            projection->effect_count += 1;
        } else if (record->record_kind == YAI_RECORD_EFFECT_RECEIPT) {
            projection->effect_count += 1;
        } else if (record->record_kind == YAI_RECORD_SUBJECT_STATE) {
            projection->subject_state_count += 1;
        } else if (record->record_kind == YAI_RECORD_RECEIPT) {
            projection->effect_count += 1;
        } else if (record->record_kind == YAI_RECORD_GRAPH_EDGE) {
            projection->graph_edge_count += 1;
        } else if (record->record_kind == YAI_RECORD_RECONSTRUCTION) {
            projection->reconstruction_count += 1;
        } else if (record->record_kind == YAI_RECORD_MEMORY_CANDIDATE) {
            projection->memory_candidate_count += 1;
            if (summary_has_memory_kind(record, "operational")) {
                projection->operational_memory_candidate_count += 1;
            } else if (summary_has_memory_kind(record, "decision")) {
                projection->decision_memory_candidate_count += 1;
            } else if (summary_has_memory_kind(record, "subject")) {
                projection->subject_memory_candidate_count += 1;
            } else if (summary_has_memory_kind(record, "error")) {
                projection->error_memory_candidate_count += 1;
            } else if (summary_has_memory_kind(record, "recovery")) {
                projection->recovery_memory_candidate_count += 1;
            }
        } else if (record->record_kind == YAI_RECORD_DIVERGENCE) {
            projection->divergence_count += 1;
            if (strstr(record->summary, "severity:critical") != 0) {
                projection->critical_divergence_count += 1;
            } else if (strstr(record->summary, "severity:warning") != 0) {
                projection->warning_divergence_count += 1;
            }
        } else if (record->record_kind == YAI_RECORD_RECONCILIATION) {
            projection->reconciliation_count += 1;
        } else if (record->record_kind == YAI_RECORD_PROJECTION_REQUEST) {
            projection->projection_request_count += 1;
            if (strstr(record->summary, "consumer:operator") != 0) {
                projection->operator_projection_count += 1;
            } else if (strstr(record->summary, "consumer:model") != 0) {
                projection->model_projection_count += 1;
            } else if (strstr(record->summary, "consumer:audit") != 0) {
                projection->audit_projection_count += 1;
            }
        } else if (record->record_kind == YAI_RECORD_PROJECTION_RESULT) {
            projection->projection_result_count += 1;
            if (strstr(record->summary, "consumer:operator") != 0) {
                projection->operator_projection_count += 1;
            } else if (strstr(record->summary, "consumer:model") != 0) {
                projection->model_projection_count += 1;
            } else if (strstr(record->summary, "consumer:audit") != 0) {
                projection->audit_projection_count += 1;
            }
            if (strstr(record->summary, "redaction:summary_only") != 0 ||
                strstr(record->summary, "redaction:refs_only") != 0 ||
                strstr(record->summary, "redaction:redacted") != 0 ||
                strstr(record->summary, "redaction:blocked") != 0) {
                projection->limited_projection_count += 1;
            }
        } else if (record->record_kind == YAI_RECORD_QUERY_RESULT) {
            projection->query_result_count += 1;
        } else if (record->record_kind == YAI_RECORD_PROJECTION_RULE) {
            projection->projection_rule_count += 1;
        } else if (record->record_kind == YAI_RECORD_AUTHORITY_SCOPE) {
            projection->authority_scope_count += 1;
        } else if (record->record_kind == YAI_RECORD_MODEL_INTERPRETATION) {
            projection->model_interpretation_count += 1;
        } else if (record->record_kind == YAI_RECORD_CASE_DOMAIN) {
            projection->case_domain_count += 1;
        } else if (record->record_kind == YAI_RECORD_CASE_ATTACHMENT) {
            projection->case_attachment_count += 1;
        } else if (record->record_kind == YAI_RECORD_CASE_BINDING) {
            projection->case_binding_count += 1;
        } else if (record->record_kind == YAI_RECORD_INTERACTION_THREAD) {
            projection->interaction_thread_count += 1;
        } else if (record->record_kind == YAI_RECORD_INTERACTION_TURN) {
            projection->interaction_turn_count += 1;
        } else if (record->record_kind == YAI_RECORD_PARTICIPANT_VIEW_FRAME) {
            projection->participant_view_frame_count += 1;
        }
    }

    (void)snprintf(projection->summary,
                   sizeof(projection->summary),
                   "projection:%s records:%zu case_domains:%zu case_attachments:%zu case_bindings:%zu interaction_threads:%zu interaction_turns:%zu participant_view_frames:%zu decisions:%zu rules:%zu projection_rules:%zu authority_scopes:%zu model_interpretations:%zu gates:%zu obligations:%zu receipt_requirements:%zu filesystem_receipts:%zu subject_states:%zu effects:%zu graph_edges:%zu reconstructions:%zu memory_candidates:%zu divergences:%zu reconciliations:%zu projection_requests:%zu projection_results:%zu query_results:%zu",
                   yai_projection_consumer_string(consumer_kind),
                   projection->source_record_count,
                   projection->case_domain_count,
                   projection->case_attachment_count,
                   projection->case_binding_count,
                   projection->interaction_thread_count,
                   projection->interaction_turn_count,
                   projection->participant_view_frame_count,
                   projection->decision_count,
                   projection->policy_rule_count,
                   projection->projection_rule_count,
                   projection->authority_scope_count,
                   projection->model_interpretation_count,
                   projection->gate_count,
                   projection->obligation_count,
                   projection->receipt_requirement_count,
                   projection->filesystem_receipt_count,
                   projection->subject_state_count,
                   projection->effect_count,
                   projection->graph_edge_count,
                   projection->reconstruction_count,
                   projection->memory_candidate_count,
                   projection->divergence_count,
                   projection->reconciliation_count,
                   projection->projection_request_count,
                   projection->projection_result_count,
                   projection->query_result_count);
    return YAI_OK;
}

const char *yai_projection_consumer_string(yai_projection_consumer_t consumer) {
    switch (consumer) {
    case YAI_PROJECTION_OPERATOR:
        return "operator";
    case YAI_PROJECTION_MODEL:
        return "model";
    case YAI_PROJECTION_AGENT:
        return "agent";
    case YAI_PROJECTION_API:
        return "api";
    case YAI_PROJECTION_AUDIT:
        return "audit";
    case YAI_PROJECTION_DEBUG:
        return "debug";
    default:
        return "unknown";
    }
}
